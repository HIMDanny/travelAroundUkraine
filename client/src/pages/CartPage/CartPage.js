import React, { useEffect } from 'react';

import { Box, Container, Stack, styled, Typography } from '@mui/material';
// Redux
import { useDispatch, useSelector } from 'react-redux';
import { fetchCart, setCartFromLocal } from '../../store/slices/cartSlice/cartSlice';

import { CartItem, TotalInfoDialog } from '../../features/Cart/components';

const ContentWrapper = styled(Stack)(({ theme }) => ({
  gap: theme.spacing(2),

  [theme.breakpoints.up('laptop')]: {
    flexDirection: 'row',
  },
}));

const CartItemsList = styled((props) => <Stack component="ul" {...props} />)(({ theme }) => ({
  padding: 0,
  margin: 0,
  flexGrow: 1,
  gap: theme.spacing(2),

  listStyleType: 'none',
}));

const CartPage = () => {
  const dispatch = useDispatch();
  const isLogin = useSelector((store) => store.user.isLogin);
  const cart = useSelector((store) => store.cart.data);

  useEffect(() => {
    isLogin ? dispatch(fetchCart()) : dispatch(setCartFromLocal());
  }, []);

  // components saved into constants
  const cartContentWrapper = (
    <ContentWrapper>
      <CartItemsList>
        {cart.map(({ product: { imageUrls, name, currentPrice, duration, itemNo, _id }, cartQuantity }) => (
          <li key={_id}>
            <CartItem
              imageUrls={imageUrls}
              name={name}
              currentPrice={currentPrice}
              duration={duration}
              cartQuantity={cartQuantity}
              itemNo={itemNo}
              id={_id}
              isLogin={isLogin}
            />
          </li>
        ))}
        <li>
          <CartItem />
        </li>
      </CartItemsList>
      <Box flexGrow={1}>
        <TotalInfoDialog cart={cart} />
      </Box>
    </ContentWrapper>
  );

  return (
    <Box marginTop="100px" marginBottom="150px" component="section" paddingY={3}>
      <Container>
        <Typography marginBottom={2} variant="h2">
          Cart
        </Typography>

        {cart.length ? (
          cartContentWrapper
        ) : (
          <Typography variant="h2" align="center">
            Your cart is empty!
          </Typography>
        )}
      </Container>
    </Box>
  );
};

export default CartPage;
