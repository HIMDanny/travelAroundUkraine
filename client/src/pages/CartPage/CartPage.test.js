/* eslint-disable react/prop-types */
import { render } from '@testing-library/react';
import CartPage from './CartPage';

jest.mock('@mui/material/Stack', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Box', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Container', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);

jest.mock('../../features/Cart/components/CartItem/CartItem', () => () => <div>Cart item</div>);
jest.mock('../../features/Cart/components/TotalInfoDialog/TotalInfoDialog', () => () => <div>Total info dialog</div>);

describe('CartPage snapshot test', () => {
  test('should CartPage match snapshot', () => {
    const { asFragment } = render(<CartPage />);

    expect(asFragment()).toMatchSnapshot();
  });
});
