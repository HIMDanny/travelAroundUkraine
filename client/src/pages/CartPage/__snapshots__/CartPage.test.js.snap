// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartPage snapshot test should CartPage match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <p>
        Cart
      </p>
      <div>
        <div>
          <li>
            <div>
              Cart item
            </div>
          </li>
          <li>
            <div>
              Cart item
            </div>
          </li>
          <li>
            <div>
              Cart item
            </div>
          </li>
        </div>
        <div>
          <div>
            Total info dialog
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
